### PR TITLE
Zendesk Destination

### PR DESCRIPTION
**Pull request by bleonard**
Closes T-368

### CODE REVIEW

File: plugins/@grouparoo/sailthru/__tests__/export/export-profile.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@ import path from "path";
 import { exportProfile } from "../../src/lib/export/exportProfile";
 import Sailthru from "../../src/lib/client";
 import { loadAppOptions, updater } from "../utils/nockHelper";
+import { helper } from "../../../../../core/api/__tests__/utils/specHelper";
```
Comment (evantahler) [inline plugins/@grouparoo/sailthru/__tests__/export/export-profile.ts:9]: I don't think this is needed?

File: core/web/pages/app/[guid]/edit.tsx [hunk 2]
```diff
@@ -309,6 +309,7 @@ export default function Page(props) {
             <Row>
               <Col>
                 {testResult.success !== null &&
+                testResult.success !== false &&
```
Comment (evantahler) [inline core/web/pages/app/[guid]/edit.tsx:312]: Good catch!

File: core/web/pages/destination/[guid]/data.tsx [hunk 3]
```diff
@@ -783,7 +783,8 @@ Page.getInitialProps = async (ctx) => {
   const { groups } = await execApi("get", `/groups`);
   const { profilePropertyRules } = await execApi(
     "get",
-    `/profilePropertyRules`
+    `/profilePropertyRules`,
+    { state: "ready" }
```
Comment (evantahler) [inline core/web/pages/destination/[guid]/data.tsx]: This appears to be on master already - https://github.com/grouparoo/grouparoo/blob/master/core/web/pages/destination/%5Bguid%5D/data.tsx#L775-L779... so what's going on git?!

File: plugins/@grouparoo/zendesk/__tests__/.env.example [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Note: These are only needed when recording new nock fixtures
+ZENDESK_SUBDOMAIN = "d3v-grouparoo"
+ZENDESK_USERNAME = "example.user@grouparoo.com"
+ZENDESK_TOKEN = "TdRrV1foBNoQ1tJRUhBsa7wJA8hONBHwV1rO87Cc"
```
Comment (evantahler) [inline plugins/@grouparoo/zendesk/__tests__/.env.example:4]: Confirming that this is random

File: plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+  },
+}) => {
+  try {
+    const client = await connect(appOptions);
+    let external_id = newProfileProperties.external_id;
+    let email = newProfileProperties.email;
+
+    const dateValue = newProfileProperties.date_field;
+
+    // console.log("exportProfile", newProfileProperties);
```
Comment (evantahler) [inline plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts]: I think these can be removed?

File: plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+      newProfileProperties,
+      oldProfileProperties
+    );
+
+    if (toDelete) {
+      if (user) {
+        // this does a soft delete
+        await client.users.delete(user.id);
+
+        // TODO: should we fully delete it? (see test for how to set it up)
```
Comment (evantahler) [inline plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts:48]: I think you made the right choice with soft-delete.  I can't think of a reason to ever /really/ delete a Zendesk profile unless there's a request to do so by the customer.

File: plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+  const out = {};
+  for (const field of required) {
+    out[field.key] = true;
+  }
+  for (const field of builtIn) {
+    out[field.key] = true;
+  }
+  return out;
+}
```
Comment (evantahler) [inline plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts:200]: ouch.

File: plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+  }
+  return "Unknown";
+}
+
+function makeTagName(groupName) {
+  // tags can't have spaces and have to be lowercase
+  let tagName = groupName.toLowerCase();
+  tagName = tagName.replace(/[^a-z]/g, "_");
+  return tagName;
+}
```
Comment (evantahler) [inline plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts:229]: this is starting to appear rather common - we could enforce this in our group names...

**Review by evantahler — COMMENTED** [review]
Awesome!

File: plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts (hunk 7 above)
Comment (bleonard) [reply plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts:200]: what's that mean? lol

File: plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts (hunk 8 above)
Comment (bleonard) [reply plugins/@grouparoo/zendesk/src/lib/export/exportProfile.ts:229]: I made a task to DRY up code that's common. and have this in it.
T-372

File: core/web/pages/destination/[guid]/data.tsx (hunk 3 above)
Comment (bleonard) [reply core/web/pages/destination/[guid]/data.tsx]: weird.

File: plugins/@grouparoo/sailthru/__tests__/export/export-profile.ts (hunk 1 above)
Comment (bleonard) [reply plugins/@grouparoo/sailthru/__tests__/export/export-profile.ts:9]: it's needed if you enable new nock so i left it there to make that easier.

File: core/web/pages/app/[guid]/edit.tsx (hunk 2 above)
Comment (bleonard) [reply core/web/pages/app/[guid]/edit.tsx:312]: feel free to make a `if(!testResult.success)` later if you want. I didn't want to mess around too much outside the scope of this PR.

**Review by evantahler — APPROVED** [review]

File: core/web/pages/destination/[guid]/data.tsx (hunk 3 above)
Comment (evantahler) [reply core/web/pages/destination/[guid]/data.tsx]: After merging, git kept what was on master... so all good!

**Comment by bleonard** [comment]
Fixes https://github.com/grouparoo/grouparoo/issues/648

**Pull request merged**